src/common/transformations/src/transformations/common_optimizations/pull_through_reduce.cpp

```diff
@@ -0,0 +1,235 @@
+bool have_same_axes(const std::vector<int64_t>& unsqueeze_axes, const std::vector<int64_t>& reduce_op_axes) {
+    return std::find_first_of(std::begin(unsqueeze_axes),
+                              std::end(unsqueeze_axes),
+                              std::begin(reduce_op_axes),
+                              std::end(reduce_op_axes)) != std::end(unsqueeze_axes);
+}
+}  // namespace
+
+ov::pass::PullUnsqueezeThroughReduce::PullUnsqueezeThroughReduce() {
+    MATCHER_SCOPE(PullUnsqueezeThroughReduce);
+
+    const auto input = pattern::any_input(pattern::has_static_rank());
+    const auto unsqueeze_axes = pattern::wrap_type<opset9::Constant>();
+    const auto unsqueeze = pattern::wrap_type<opset9::Unsqueeze>({input, unsqueeze_axes}, pattern::has_static_rank());
+    const auto reduce_axes = pattern::wrap_type<opset9::Constant>();
+    const auto arithmetic_reduce = pattern::wrap_type<op::util::ArithmeticReductionKeepDims>({unsqueeze, reduce_axes});
+    const auto logical_reduce = pattern::wrap_type<op::util::LogicalReductionKeepDims>({unsqueeze, reduce_axes});
+    const auto reduce = std::make_shared<pattern::op::Or>(OutputVector{arithmetic_reduce, logical_reduce});
+
+    matcher_pass_callback callback = [=](pattern::Matcher& m) {
+        auto& pattern_map = m.get_pattern_value_map();
+        const auto input_node = pattern_map.at(input);
+        const auto reduce_node = pattern_map.find(arithmetic_reduce) != std::end(pattern_map)
+                                     ? pattern_map.at(arithmetic_reduce).get_node_shared_ptr()
+                                     : pattern_map.at(logical_reduce).get_node_shared_ptr();
+        const auto unsqueeze_node = pattern_map.at(unsqueeze).get_node_shared_ptr();
+        auto unsqueeze_axes_input =
+            std::dynamic_pointer_cast<opset9::Constant>(pattern_map.at(unsqueeze_axes).get_node_shared_ptr());
+        auto reduce_axes_input =
+            std::dynamic_pointer_cast<opset9::Constant>(pattern_map.at(reduce_axes).get_node_shared_ptr());
+
+        auto unsqueeze_axes_val = unsqueeze_axes_input->cast_vector<int64_t>();
+        normalize_axes(unsqueeze_node.get(),
+                       unsqueeze_node->get_output_partial_shape(0).rank().get_length(),
+                       unsqueeze_axes_val);
+        const auto reduce_axes_val = std::dynamic_pointer_cast<op::util::ArithmeticReductionKeepDims>(reduce_node)
+                                         ? std::dynamic_pointer_cast<op::util::ArithmeticReductionKeepDims>(reduce_node)
+                                               ->get_reduction_axes()
+                                               .to_vector()
+                                         : std::dynamic_pointer_cast<op::util::LogicalReductionKeepDims>(reduce_node)
+                                               ->get_reduction_axes()
+                                               .to_vector();
```

Comment:
Please make `get_reduction_axes` a method of ReductionBase with implementation from op/util/arithmetic_reduction.cpp

Reply:
Unfortunately, `ReductionBase` has no `get_reduction_axes`; `ArithmeticReductionKeepDims` and `LogicalReductionKeepDims` have separate impl

Reply:
I asked to introduce such a method... but okay I guess.

Reply:
I agree with @jane-intel it's better to move common method `get_reduction_axes` to `ReductionBase` class to avoid these dynamic_pointer_casts.

Reply:
done